text/0031-configuration.md

```diff
@@ -0,0 +1,287 @@
+- have getters or `pub(crate)` fields;
+
+Here is a small example fitting all these requirements:
+
+`config.toml`
+```toml
+[snapshot]
+meta_file_path  = "./data/snapshot/mainnet.snapshot.meta"
+state_file_path = "./data/snapshot/mainnet.snapshot.state"
+```
+
+`config.rs`
+```rust
+#[derive(Clone)]
+pub struct SnapshotConfig {
+    meta_file_path: String,
+    state_file_path: String,
+}
+
+impl SnapshotConfig {
+    pub fn build() -> SnapshotConfigBuilder {
+      SnapshotConfig::new()
+    }
+
+    pub fn meta_file_path(&self) -> &String {
+        &self.meta_file_path
+    }
+
+    pub fn state_file_path(&self) -> &String {
+        &self.state_file_path
+    }
+}
+
+```
+
+### Read a configuration builder from a file
+
+```rust
+let config_builder = match fs::read_to_string("config.toml") {
+    Ok(toml) => match toml::from_str::<SnapshotConfigBuilder>(&toml) {
+        Ok(config_builder) => config_builder,
+        Err(e) => {
+            panic!("Error parsing config file: {:?}", e);
+        }
+    },
+    Err(e) => {
+        panic!("Error reading config file: {:?}", e);
+    }
+};
+
+// Override fields if necessary e.g. with CLI arguments.
+
+let config = config_builder.finish();
+```
+
+### Write a configuration to a file
+
+```rust
+match toml::to_string(&config) {
+    Ok(toml) => match fs::File::create("config.toml") {
+        Ok(mut file) => {
+            if let Err(e) = file.write_all(toml.as_bytes()) {
+                panic!("Error writing .toml config file: {:?}", e);
+            }
+        }
+        Err(e) => {
+            panic!("Error creating .toml config file: {:?}", e);
+        }
+    },
+    Err(e) => {
+        panic!("Error serializing .toml config file: {:?}", e);
+    }
+}
+```
+
+### Sub-configuration
+
+It is also very easy to create sub-configurations by nesting configuration builders and configurations.
+
+`config.toml`
+```toml
+[snapshot]
+[snapshot.local]
+meta_file_path  = "./data/snapshot/local/mainnet.snapshot.meta"
+state_file_path = "./data/snapshot/local/mainnet.snapshot.state"
+[snapshot.global]
+file_path       = "./data/snapshot/global/mainnet.txt"
+```
+
+`config.rs`
+```rust
+#[derive(Default, Deserialize)]
+pub struct LocalSnapshotConfigBuilder {
+    meta_file_path: Option<String>,
+    state_file_path: Option<String>,
+}
+
+#[derive(Default, Deserialize)]
+pub struct GlobalSnapshotConfigBuilder {
+    file_path: Option<String>,
+}
+
+#[derive(Default, Deserialize)]
+pub struct SnapshotConfigBuilder {
+    local: LocalSnapshotConfigBuilder,
+    global: GlobalSnapshotConfigBuilder,
+}
+
+impl SnapshotConfigBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    // Setters
+
+    pub fn finish(self) -> SnapshotConfig {
+        SnapshotConfig {
+            local: LocalSnapshotConfig {
+                meta_file_path: self
+                    .local
+                    .meta_file_path
+                    .unwrap_or(DEFAULT_LOCAL_SNAPSHOT_META_FILE_PATH.to_string()),
```

Comment:
Wouldn't it be better to build those types around `str` instead of `String`. Then we wouldn't need those allocations here ^.

Reply:
It can't be `String` since, when not default, it comes from the user.

Reply:
alright :+1: